### PR TITLE
[SPARK-26432][CORE] Obtain HBase delegation token operation compatible with HBase 2.x.x version API 

### PR DESCRIPTION
**Pull request by sujith71955**
## What changes were proposed in this pull request?

While obtaining token from hbase service , spark uses deprecated API of hbase ,
```public static Token<AuthenticationTokenIdentifier> obtainToken(Configuration conf)```
This deprecated API is already been removed from hbase 2.x version as part of the hbase 2.x major release. https://issues.apache.org/jira/browse/HBASE-14713_
there is one more stable API in 
```public static Token<AuthenticationTokenIdentifier> obtainToken(Connection conn)``` in TokenUtil class
spark shall use this stable api for getting the delegation token.

To invoke this api first connection object has to be retrieved from ConnectionFactory and the same connection can be passed to obtainToken(Connection conn) for getting token.
eg: Call   ```public static Connection createConnection(Configuration conf)```
, then call   ```public static Token<AuthenticationTokenIdentifier> obtainToken( Connection conn)```.

## How was this patch tested?
Manual testing is been done.
Manual test result:
Before fix:

hbase-dep-obtaintok 1

After fix:
1. Create 2 tables in hbase shell
 >Launch hbase shell
 >Enter commands to create tables and load data
    create 'table1','cf'
    put 'table1','row1','cf:cid','20'

    create 'table2','cf'
    put 'table2','row1','cf:cid','30'
	
 >Show values command
   get 'table1','row1','cf:cid'  will diplay value as 20
   get 'table2','row1','cf:cid'  will diplay value as 30
   
   
2.Run SparkHbasetoHbase class in testSpark.jar using spark-submit

spark-submit --master yarn-cluster --class com.mrs.example.spark.SparkHbasetoHbase --conf "spark.yarn.security.credentials.hbase.enabled"="true" --conf "spark.security.credentials.hbase.enabled"="true" --keytab /opt/client/user.keytab --principal sen testSpark.jar

The SparkHbasetoHbase test class will update the value of table2 with sum of values of table1 & table2.

table2 = table1+table2
As we can see in the snapshot the spark job has been successfully able to interact with hbase service and able to update the row count.
obtaintok_success 1

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
ok to test

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala [hunk 1]
```diff
@@ -39,23 +41,46 @@ private[security] class HBaseDelegationTokenProvider
...
-      logDebug("Attempting to fetch HBase security token.")
-      val token = obtainToken.invoke(null, hbaseConf(hadoopConf))
-        .asInstanceOf[Token[_ <: TokenIdentifier]]
-      logInfo(s"Get token from HBase: ${token.toString}")
-      creds.addToken(token.getService, token)
-    } catch {
-      case NonFatal(e) =>
-        logWarning(s"Failed to get token from service $serviceName", e)
+      // Token<AuthenticationTokenIdentifier> obtainToken(Configuration conf) is a deprecated
+      // method and in Hbase 2.1.1 the method is already removed. there is one more public
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: Could you be more specific? Is this removed at 2.1.1?

Comment (sujith71955) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: shall we mention as 2.x, they removed it as part of Hbase 2.0.0 major version upgrade

Comment (sujith71955) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: https://issues.apache.org/jira/browse/HBASE-14713_ - JIRA for reference

Comment (sujith71955) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: updated the version. thnks .

**Comment by vanzin** [comment]
Does your code keep support for HBase 1.x? Looks like it does not; that's the main reason why I just didn't submit my patch for this:
https://github.com/cloudera/spark/commit/3de80e469211797a099798e8bbe2b1fba74ff3df#diff-6e15cee083da1fefc4b9fa8a1f8642e9

It should be easy to keep both by just keeping the existing code in a different method, can calling that one if the 2.x method fails.

Also, please use the PR title to explain the solution, not the problem. Problems are described in the bug, not the PR.

**Comment by sujith71955** [comment]
> HBase 1.x

The API which currently i am using is introduced in

**Comment by sujith71955** [comment]
> Does your code keep support for HBase 1.x? Looks like it does not; that's the main reason why I just didn't submit my patch for this:
> [cloudera/spark@3de80e4#diff-6e15cee083da1fefc4b9fa8a1f8642e9](https://github.com/cloudera/spark/commit/3de80e469211797a099798e8bbe2b1fba74ff3df#diff-6e15cee083da1fefc4b9fa8a1f8642e9)
> 
> It should be easy to keep both by just keeping the existing code in a different method, then calling that one if the 2.x method fails.
> 
> Also, please use the PR title to explain the solution, not the problem. Problems are described in the bug, not the PR.

Will update based on your comment. Thanks for the valuable feedback.

**Comment by vanzin** [comment]
> The API which currently i am using obtainToken(Connection conn) is introduced in Hbase 0.98.9

Good to know. When I find some time I'll give this a try, just in case... it's hard to trust in reflection code without testing. :-/

**Comment by sujith71955** [comment]
> > The API which currently i am using obtainToken(Connection conn) is introduced in Hbase 0.98.9
> 
> Good to know. When I find some time I'll give this a try, just in case... it's hard to trust in reflection code without testing. :-/

Definitely you can give a try, anyways i will test and update the report in the PR, i already tested it with my current code and its working fine. Need to retest with  modified code. Thnks

**Comment by vanzin** [comment]
Could you also remove the WIP from the title? Unless you don't actually want this to be pushed in the current form.

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala [hunk 2]
```diff
@@ -50,12 +53,70 @@ private[security] class HBaseDelegationTokenProvider
       creds.addToken(token.getService, token)
     } catch {
       case NonFatal(e) =>
-        logWarning(s"Failed to get token from service $serviceName", e)
+        logWarning(s"Failed to get token from service $serviceName" + "Trying to invoke" +
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: `$serviceName" + "Trying` -> `$serviceName. Trying`?

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala [hunk 3]
```diff
@@ -50,12 +53,70 @@ private[security] class HBaseDelegationTokenProvider
...
+        // another compatible API of HBase service.
+        obtainDelegationTokensWithHBaseConn(hadoopConf, creds)
     }
-
     None
   }
 
+  /**
+   * The HBase client API used in below method is introduced from HBase 0.98.9,1.0.0 version
+    * to invokethis api first connection object has to be retrieved from ConnectionFactory and the
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: indentation.

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala [hunk 4]
```diff
@@ -50,12 +53,70 @@ private[security] class HBaseDelegationTokenProvider
...
+    * same connection can be passed to
+    * Token<AuthenticationTokenIdentifier> obtainToken(Connection conn) API
+    *
+    * @param hadoopConf
+    * @param creds
+    */
+  def obtainDelegationTokensWithHBaseConn(hadoopConf: Configuration, creds: Credentials): Unit = {
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: `private def`?

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala (hunk 3 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: `invokethis` -> `invoke this`

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala [hunk 5]
```diff
@@ -50,12 +53,70 @@ private[security] class HBaseDelegationTokenProvider
...
+    catch {
+      case NonFatal(e) =>
+        logError(s"Failed to get token from service $serviceName", e)
+    }
+    finally {
+      conn.asInstanceOf[Closeable].close()
+    }
+  }
+
+  def getClassParamType(arrayOfMethod: Array[Method]): Class[_] = {
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: `private def`?

Comment (sujith71955) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: yes, should be

Comment (sujith71955) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: done. updated. thnks

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala (hunk 2 above)
Comment (sujith71955) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: handled, let me know if any more corrections required. thnks

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala (hunk 4 above)
Comment (sujith71955) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: done

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala [hunk 6]
```diff
@@ -50,12 +53,71 @@ private[security] class HBaseDelegationTokenProvider
...
+      case NonFatal(e) =>
+        logError(s"Failed to get token from service $serviceName", e)
+    }
+    finally {
+      conn.asInstanceOf[Closeable].close()
+    }
+  }
+
+  private def getClassParamType(arrayOfMethod: Array[Method]): Class[_] = {
+    val seqOfFiltMethod = arrayOfMethod.filter(f => f.getName.equalsIgnoreCase("obtainToken"))
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: Nits: just write `filter(_.getName....)`. You can't ignore case here, right?

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala [hunk 7]
```diff
@@ -50,12 +53,71 @@ private[security] class HBaseDelegationTokenProvider
...
+      getMethod("createConnection", classOf[Configuration])
+    val conn = obtainHBaseConnection.invoke(null, hbaseConf(hadoopConf))
+    try {
+      val arrayOfMethod: Array[Method] = ClassLoader.getSystemClassLoader
+        .loadClass("org.apache.hadoop.hbase.security.token.TokenUtil").getMethods
+      val obtainParamTypeClass = getClassParamType(arrayOfMethod)
+      if (null == obtainParamTypeClass) {
+        logDebug(s"Failed to invoke service to get token from service $serviceName")
+      }
+      else {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: Nit: pull these up onto the previous line. Same for catch below.

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala [hunk 8]
```diff
@@ -50,12 +53,71 @@ private[security] class HBaseDelegationTokenProvider
...
+      hadoopConf: Configuration, creds: Credentials): Unit = {
+    val mirror = universe.runtimeMirror(Utils.getContextOrSparkClassLoader)
+    // Token<AuthenticationTokenIdentifier> obtainToken(Configuration conf) is a deprecated
+    // method and in Hbase 2.0.0 the method is already removed. there is one more public
+    // consistent API Token<AuthenticationTokenIdentifier> obtainToken(Connection conn),
+    // in TokenUtil class which can be invoked using reflection.
+    val obtainHBaseConnection = mirror.classLoader.
+      loadClass("org.apache.hadoop.hbase.client.ConnectionFactory").
+      getMethod("createConnection", classOf[Configuration])
+    val conn = obtainHBaseConnection.invoke(null, hbaseConf(hadoopConf))
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: Can you cast this to at least `Closeable`? you do later. Is it known to be something more specific or does the type vary with version?

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala [hunk 9]
```diff
@@ -50,12 +53,71 @@ private[security] class HBaseDelegationTokenProvider
...
+   * @param creds
+   */
+  private def obtainDelegationTokensWithHBaseConn(
+      hadoopConf: Configuration, creds: Credentials): Unit = {
+    val mirror = universe.runtimeMirror(Utils.getContextOrSparkClassLoader)
+    // Token<AuthenticationTokenIdentifier> obtainToken(Configuration conf) is a deprecated
+    // method and in Hbase 2.0.0 the method is already removed. there is one more public
+    // consistent API Token<AuthenticationTokenIdentifier> obtainToken(Connection conn),
+    // in TokenUtil class which can be invoked using reflection.
+    val obtainHBaseConnection = mirror.classLoader.
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: What if these throw exceptions? later reflective access is wrapped in try-catch; is this different?

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala [hunk 10]
```diff
@@ -50,12 +53,71 @@ private[security] class HBaseDelegationTokenProvider
...
+      loadClass("org.apache.hadoop.hbase.client.ConnectionFactory").
+      getMethod("createConnection", classOf[Configuration])
+    val conn = obtainHBaseConnection.invoke(null, hbaseConf(hadoopConf))
+    try {
+      val arrayOfMethod: Array[Method] = ClassLoader.getSystemClassLoader
+        .loadClass("org.apache.hadoop.hbase.security.token.TokenUtil").getMethods
+      val obtainParamTypeClass = getClassParamType(arrayOfMethod)
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: Could we rename the vars and methods here? I'm not clear what they do as they're named pretty generically. Are you looking for a method with a specific signature? the reflection API has methods to do that directly.

**Review by srowen — CHANGES_REQUESTED** [review]

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala (hunk 6 above)
Comment (sujith71955) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: This Method is not required, as you told from reflection we can directly load the Class of the Connection  as below, we can omit the above logic.
  val connectionParamTypeClassRef = mirror.classLoader
        .loadClass("org.apache.hadoop.hbase.client.Connection")

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala (hunk 7 above)
Comment (sujith71955) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: done

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala (hunk 8 above)
Comment (sujith71955) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: done

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala (hunk 9 above)
Comment (sujith71955) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: i had pulled down the statement inside try catch block, declared the connection variable above the block for later access.

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala (hunk 10 above)
Comment (sujith71955) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: Logic not required now, remaining variable i renamed to more meaningful name, any suggestions welcome. thanks

**Comment by sujith71955** [comment]
@srowen @dongjoon-hyun @vanzin  Handled the comments and tested . thanks for the valuable suggestions. let me know for any clarifications or suggestions

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala [hunk 11]
```diff
@@ -50,12 +52,59 @@ private[security] class HBaseDelegationTokenProvider
...
     None
   }
 
+  /**
+   * The HBase client API used in below method is introduced from HBase 0.98.9,1.0.0 version
+   * to invoke this api first connection object has to be retrieved from ConnectionFactory and the
+   * same connection can be passed to
+   * Token<AuthenticationTokenIdentifier> obtainToken(Connection conn) API
+   *
+   * @param hadoopConf
```
Comment (gaborgsomogyi) [inline core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: I would either add real documentation or drop the params.

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala [hunk 12]
```diff
@@ -50,12 +52,59 @@ private[security] class HBaseDelegationTokenProvider
...
+   * @param creds
+   */
+  private def obtainDelegationTokensWithHBaseConn(
+      hadoopConf: Configuration, creds: Credentials): Unit = {
+    val mirror = universe.runtimeMirror(Utils.getContextOrSparkClassLoader)
```
Comment (gaborgsomogyi) [inline core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: Why not having this under try umbrella?

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala [hunk 13]
```diff
@@ -50,12 +52,59 @@ private[security] class HBaseDelegationTokenProvider
       creds.addToken(token.getService, token)
     } catch {
       case NonFatal(e) =>
-        logWarning(s"Failed to get token from service $serviceName", e)
+        logWarning(s"Failed to get token from service $serviceName. " +
+          s"Retrying to invoke service with hbase connection.")
+        // Seems to be spark is trying to get the token from HBase 2.x.x  version or above where the
+        // obtainToken(Configuration conf) API is been removed. Lets try obtaining the token from
+        // another compatible API of HBase service.
+        obtainDelegationTokensWithHBaseConn(hadoopConf, creds)
```
Comment (gaborgsomogyi) [inline core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: Wondering whether it would be better to try with the stable API first. Later it would be easier to removed 1.x support.

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala [hunk 14]
```diff
@@ -50,12 +52,59 @@ private[security] class HBaseDelegationTokenProvider
       creds.addToken(token.getService, token)
     } catch {
       case NonFatal(e) =>
-        logWarning(s"Failed to get token from service $serviceName", e)
```
Comment (gaborgsomogyi) [inline core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: Why dropped exception printing? It could still hold useful information if 2.x also failed.
Additionally wondering whether it's really a warning?

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala [hunk 15]
```diff
@@ -50,12 +52,59 @@ private[security] class HBaseDelegationTokenProvider
...
+        loadClass("org.apache.hadoop.hbase.security.token.TokenUtil").
+        getMethod("obtainToken", connectionParamTypeClassRef)
+      logDebug("Attempting to fetch HBase security token.")
+      val token = obtainTokenMethod.invoke(null, connectionHbaseClient)
+        .asInstanceOf[Token[_ <: TokenIdentifier]]
+      logInfo(s"Get token from HBase: ${token.toString}")
+      creds.addToken(token.getService, token)
+    } catch {
+      case NonFatal(e) =>
+        logError(s"Failed to get token from service $serviceName", e)
```
Comment (gaborgsomogyi) [inline core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: The original level was warning, why error here?

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala (hunk 13 above)
Comment (sujith71955) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: I think both way shall not make much difference, still we can take a call by getting opinion from other reviewers. thnks.

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala (hunk 11 above)
Comment (sujith71955) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: updated the comment. thnks.

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala (hunk 12 above)
Comment (sujith71955) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: right, we can move the statement inside. done.

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala (hunk 15 above)
Comment (sujith71955) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: I thought both the attempts of retrieving delegation token got failed, so basically spark will not able interact with hbase service in this scenario which will in-turn fail the user  attempt to do some operations based on hbase. so i thought better to log this as error instead of just warning. in  the scenario warning user can still successfully do his operations with some limitations at max. anyways i updated the level to error as the old code also was logging this failure as warning. Thanks.

**Comment by sujith71955** [comment]
@gaborgsomogyi thanks for the valuable feedback, let me know for any clarification you have still.
@vanzin @dongjoon-hyun @srowen  Please let me know if the above modifications are fine, or still any improvements required for the PR. thanks all for your valuable time.

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala (hunk 8 above)
Comment (sujith71955) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: Yes, we can cast, no type shall not vary with version .

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala (hunk 14 above)
Comment (sujith71955) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: yes better to convey the error to  user, i modified the message . thnks

**Comment by sujith71955** [comment]
@all 
Please let me know is any issues still needs to be addressed in this PR,  Without this fix  spark cannot be connected to hbase 2.x version which seems to be a major problem for hbase related usecases.

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala [hunk 16]
```diff
@@ -50,12 +52,59 @@ private[security] class HBaseDelegationTokenProvider
       creds.addToken(token.getService, token)
     } catch {
       case NonFatal(e) =>
-        logWarning(s"Failed to get token from service $serviceName", e)
+        logWarning(s"Failed to get token from service $serviceName due to  " + e +
+          s" Retrying to fetch HBase security token with hbase connection parameter.")
+        // Seems to be spark is trying to get the token from HBase 2.x.x  version or above where the
+        // obtainToken(Configuration conf) API is been removed. Lets try obtaining the token from
```
Comment (gaborgsomogyi) [inline core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: Nit: is -> has

**Review by gaborgsomogyi — COMMENTED** [review]
Other than the small thing looks good.
Maybe it's better not to use force push all the time because it's hard to follow the PR.

**Comment by sujith71955** [comment]
> Other than the small thing looks good.
> Maybe it's better not to use force push all the time because it's hard to follow the PR.

Valid point, will take care in future. thanks for suggestion

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala (hunk 16 above)
Comment (sujith71955) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: fixed. thanks

**Review by gaborgsomogyi — APPROVED** [review]

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala [hunk 17]
```diff
@@ -50,12 +52,59 @@ private[security] class HBaseDelegationTokenProvider
...
+        // obtainToken(Configuration conf) API has been removed. Lets try obtaining the token from
+        // another compatible API of HBase service.
+        obtainDelegationTokensWithHBaseConn(hadoopConf, creds)
     }
-
     None
   }
 
+  /**
+   * The HBase client API used in below method is introduced from HBase 0.98.9,1.0.0 version
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: 0.98.9 or 1.0.0? Pick one.

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala [hunk 18]
```diff
@@ -50,12 +52,59 @@ private[security] class HBaseDelegationTokenProvider
...
+   * to invoke this api first connection object has to be retrieved from ConnectionFactory and the
+   * same connection can be passed to
+   * Token<AuthenticationTokenIdentifier> obtainToken(Connection conn) API
+   *
+   * @param hadoopConf Configuration of current Hadoop Compatible system.
+   * @param creds Credentials to add tokens and security keys to.
+   */
+  private def obtainDelegationTokensWithHBaseConn(
+      hadoopConf: Configuration, creds: Credentials): Unit = {
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: One argument per line.

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala [hunk 19]
```diff
@@ -50,12 +52,59 @@ private[security] class HBaseDelegationTokenProvider
...
+    var connectionHbaseClient = null.asInstanceOf[Closeable]
+    try {
+      val mirror = universe.runtimeMirror(Utils.getContextOrSparkClassLoader)
+      // Token<AuthenticationTokenIdentifier> obtainToken(Configuration conf) is a deprecated
+      // method and in Hbase 2.0.0 the method is already removed. there is one more public
+      // consistent API Token<AuthenticationTokenIdentifier> obtainToken(Connection conn),
+      // in TokenUtil class which can be invoked using reflection.
+      val connectionFactoryClass = mirror.classLoader.
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: Move all `.` to the next line.

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala [hunk 20]
```diff
@@ -50,12 +52,59 @@ private[security] class HBaseDelegationTokenProvider
...
+   *
+   * @param hadoopConf Configuration of current Hadoop Compatible system.
+   * @param creds Credentials to add tokens and security keys to.
+   */
+  private def obtainDelegationTokensWithHBaseConn(
+      hadoopConf: Configuration, creds: Credentials): Unit = {
+    var connectionHbaseClient = null.asInstanceOf[Closeable]
+    try {
+      val mirror = universe.runtimeMirror(Utils.getContextOrSparkClassLoader)
+      // Token<AuthenticationTokenIdentifier> obtainToken(Configuration conf) is a deprecated
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: Already explained in the method comment.

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala [hunk 21]
```diff
@@ -50,12 +52,59 @@ private[security] class HBaseDelegationTokenProvider
...
+   * to invoke this api first connection object has to be retrieved from ConnectionFactory and the
+   * same connection can be passed to
+   * Token<AuthenticationTokenIdentifier> obtainToken(Connection conn) API
+   *
+   * @param hadoopConf Configuration of current Hadoop Compatible system.
+   * @param creds Credentials to add tokens and security keys to.
+   */
+  private def obtainDelegationTokensWithHBaseConn(
+      hadoopConf: Configuration, creds: Credentials): Unit = {
+    var connectionHbaseClient = null.asInstanceOf[Closeable]
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: `var connectionHbaseClient: Closeable = null`

Also `hbaseConnection` or just `connection` is a better name.

**Review by vanzin — COMMENTED** [review]
I tested this on old hbase and it works; but I thought you were going to keep only the new code, after you said that it was compatible with old hbase.

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala (hunk 18 above)
Comment (sujith71955) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: updated. thanks

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala (hunk 21 above)
Comment (sujith71955) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: updated. thanks

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala (hunk 20 above)
Comment (sujith71955) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: updated. thanks

File: core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala (hunk 19 above)
Comment (sujith71955) [reply core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala]: done. thanks

**Comment by sujith71955** [comment]
> I tested this on old hbase and it works; but I thought you were going to keep only the new code, after you said that it was compatible with old hbase.

Thanks a lot for testing vanzin, even i tested and attached the report above please refer.
I updated based on the below remarks, and i think it make sense as this particular API is been  introduced only  in 0.98.9 version. for older versions this approach will help us.Hope its fine, let me know for any suggestions.
image

**Comment by sujith71955** [comment]
> Test FAILed.
> Refer to this link for build results (access rights to CI server needed):
> https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7358/
> Test FAILed.

Seems to be unknown error, can someone re-trigger the build. thanks

**Comment by gaborgsomogyi** [comment]
retest this please

**Comment by vanzin** [comment]
looks good pending tests (which really shouldn't be affected anyway...)

**Comment by vanzin** [comment]
retest this please

**Comment by vanzin** [comment]
Merging to master.
